pkg/yang/entry_test.go

```diff
@@ -1412,3 +1412,74 @@ func TestEntryTypes(t *testing.T) {
 		}
 	}
 }
+
+func TestSequenceNumber(t *testing.T) {
+	getdir := func(e *Entry, elements ...string) (*Entry, error) {
+		for _, elem := range elements {
+			next := e.Dir[elem]
+			if next == nil {
+				return nil, fmt.Errorf("%s missing directory %q", e.Path(), elem)
+			}
+			e = next
+		}
+		return e, nil
+	}
+
+	modtext := `
+module sequence {
```

Comment:
Thanks for adding the test; could you please extend it (either another case or modify this case is fine by me) to cover grouping and uses?  Ones defined in the same module are fine for the test.

I'm guessing, but a useful test for this would be to define a uses inside of an "rpc" or "action" statement, because that's the (only?) place, where sequence order matters.

So, I imagine something along the lines of;

```
module sequence {
  ...
  grouping testGroup1 {
    leaf foo1 { type string; }
    uses testGroup2;
    leaf bar1 { type string; }
  } 

  grouping testGroup1 {
    leaf foo2 { type string; }
    leaf bar2 { type string; }
  } 

  rpc "sequence" {
    input {
      leaf id { type string; }
      uses testGroup2;
      leaf tail { type string; }
    }
  }
```

We'd expect the `SequenceNum` within the `input` statement to be:

0 -> `id`
1 -> `foo1`
2 -> `foo2`
3 -> `bar2`
4 -> `bar1`
5 -> `tail`

For extra points, another case to test would be when an `augment` is applied to the `*yang.Entry` in question.  That's about all the expressions of `(*yang.Entry).add()` I can think of.

Reply:
correction: of course, in the `module sequence` above, the second instance of `grouping testGroup1` should read `grouping testGroup2`.

Reply:
And the uses statement contradicts the expected output.  Apologies for the bad example, but I imagine you get the idea; we want to ensure that when the `uses` statement is scanned by `ToEntry`, we'll see the grouping being injected in the correct ordering.

Reply:
Appreciate the help here, figured it wouldn't be quite so easy. Looks like the biggest problem is group expansion, we lose ordering when that happens (groups are also merged into the parent node before other leaves). 

Instead, taking another approach of trying to resolve the groups and recover the ordering from the AST appears to work in the nested group case. This is also possible from external modules:

```go
func findOrderedChildren(e yang.Node, seen map[string]bool) []string {
	res := []string{}
	for _, stmt := range e.Statement().SubStatements() {
		// If it is a uses statement, and we can resolve the group recurse into it
		if stmt.Kind() == (&yang.Uses{}).Kind() {
			if grp := yang.FindGrouping(e, stmt.NName(), seen); grp != nil {
				res = append(res, findOrderedChildren(grp, seen)...)
				continue
			}
		}

		res = append(res, stmt.NName())
	}
	return res
}
```

I know there are corner cases this misses, I'm not sure how to get augments (not sure how to associate entries coming from an augment back to their augment).

Reply:
Yeah, the above will miss augments.  Don't forget `choice` (and then `case`) statements, also - their children data nodes are naturally part of their parent schema node (aka parent `*Entry`) and will be relevant to answering questions about child data definitions at that node.

Augments cause direct changes to the `*yang.Entry` in their targets, so once they've been applied there's (currently) no straight-forward way to say that a `*yang.Entry` was the result of an augmentation, though you could inspect its `.Node` to see where it came from (it's normally but not required to be a different module, so scanning Parent-bound from the augmented Entry's Node will eventually hit an "augment" statement).

Other than that, the only "easy" change I can imagine now without major refactoring is to store child entries in an additional slice managed by `(*yang.Entry).add()`. You'd then iterate over the slice, considering `choice` and `case` statements you found along the way as relevant to your use case.

That... may be simple, again the challenge is in the confirmation.  Places which would have to be considered include `(*yang.Entry).FixChoice()` and `(*yang.Entry).merge()`, in particular, along with `add()`.